Implemented Comprehensive Development Environment Setup and Enhanced Contribution Guidelines (#57)
shardeum/shardeum: README.md

## Table of Contents

1. [Introduction](#introduction)
2. [Features](#features)
3. [Community](#community)
4. [Getting Started](#getting-started)
    - [Prerequisites](#prerequisites)
    - [Installation](#installation)
5. [Usage](#usage)
    - [Running the Node](#running-the-node)
    - [Interacting with the Blockchain](#interacting-with-the-blockchain)
6. [Local Development](#local-development)
    - [Environment Setup](#environment-setup)
    - [Node.js and NVM](#nodejs-and-nvm)
    - [Rust Toolchain](#rust-toolchain)
    - [Build Essentials](#build-essentials)
    - [Project Initialization](#project-initialization)
7. [Configuration](#configuration)
    - [Easy Mode](#easy-mode)
    - [Advanced Mode](#advanced-mode)
8. [Running the Network](#running-the-network)
9. [Stopping the Network and Cleanup](#stopping-the-network-and-cleanup)
10. [Testing and Metamask](#testing-and-metamask)
    - [Install Metamask](#install-metamask)
    - [Adding Shardeum Network to Metamask](#adding-shardeum-network-to-metamask)
11. [Starting the JSON-RPC Server](#starting-the-json-rpc-server)
12. [Contributing](#contributing)
13. [License](#license)
14. [Contact](#contact)

## Introduction

> Shardeum is built to solve the scalability and performance challenges faced by existing blockchain networks. By implementing sharding, Shardeum ensures that the network can handle a large number of transactions per second without compromising security or decentralization.

Reviewer: ```suggestion
Shardeum is an EVM-compliant autoscaling blockchain platform that leverages dynamic state sharding to enhance scalability of the network. By implementing a sharding model, Shardeum ensures that the network can offer faster processing times and lower transaction costs without compromising security or decentralization.
```
I'm suggesting this update so we can keep our definition of Shardeum a bit more consistent with our existing social messaging

[Join the Shardeum Discord Server](https://discord.com/invite/shardeum)

## Getting Started

### Prerequisites

Before you begin, ensure you have the following installed on your system:

- Node.js (version 18.16.1)
- npm (version 9.5.1)
- Docker (optional, for containerized deployment)

### Installation

1. **Clone the Repository**:
    ```bash
    git clone https://github.com/shardeum/shardeum.git
    cd shardeum
    ```

2. **Install Dependencies**:
    ```bash
    npm install
    ```

> 3. **Setup Environment Variables**:

Reviewer: Since we do not enforce any environment variable config before running the network, it might be better to remove this step. It'll save some development time, so the faster we can get to a successfully running network the better :)

    Copy the `.env_example` to `.env` and customize the settings as needed.
    ```bash
    cp .env_example .env
    ```

## Usage

> ### Running the Node

Reviewer: I'll suggest moving this section below the `Environment Setup` up section, otherwise the npm start command would fail and devs may not know how to debug.

> 1. **Start the Node**:

Reviewer: I would suggest removing this section as we already have it covered under the local Development section. Also, If they npm run start at this point, the command might fail and devs may not know how to debug.

    ```bash
    npm run start
    ```

2. **Run in Docker (Optional)**:
    ```bash
    docker build -t shardeum .
    docker run -d -p 8080:8080 shardeum
    ```

> ### Interacting with the Blockchain

Reviewer: We can also remove this section in the interest of getting developers set up as quickly as possible. We don't also have a public API reference yet

npm config list                         // Verify the configuration
```

> ### Downloading Project Source Code

Reviewer: ```suggestion
### Download Project Source Code
```

Now that the environment is setup, let’s install the project source code and switch to `dev` branch:

```bash
git clone https://github.com/shardeum/shardeum.git
cd server
git switch dev
```

> #### Installing npm Dependencies

Reviewer: ```suggestion
#### Install npm Dependencies
```

Once you are on the `dev` branch, install the local node modules to build the project by the following command:

```bash
npm ci
```

### Project Initialization

Now, you can initialize the project using the following command:

```bash
npm run prepare
```

> ### Installing the Shardus CLI

Reviewer: ```suggestion
### Install the Shardus CLI
```

[Shardus](https://docs.shardus.com/docs/tools/shardus-cli-tool/README) is a command line tool from Shardus SDK that makes launching any Shardus protocol powered network easier:

```bash
npm install -g shardus
npm update @shardus/archiver
```

## Configuration

> You have two ways to configure your local environment: easy mode or advanced mode.

Reviewer: ```suggestion
There are two ways to configure your local environment, automatic mode and manual mode.
```

I'm opting for automatic/manual instead of easy/advanced because it better communicates what is going on and also avoids giving the impression that one mode is superior to the other.

txBalancePreCheck: false,
StakingEnabled: false
```

3. **Local Testing:** Additionally, for local testing, adjust the following settings. These changes allow nodes to get active in 30 seconds compared to 60 seconds and generate a new block every 3 seconds instead of 6 seconds, enhancing your local testing efficiency.

```typescript
// src/config/index.ts
cycleDuration: 30,

// src/shardeum/shardeumFlags.ts
blockProductionRate: 3,
```

After completing the above steps, compile again with `npm run prepare`.

## Running the Network

Now we’re ready to create the network from within the Shardeum validator repo:

```bash
shardus start 20
```

This command will start a Shardeum network of 20 nodes, 1 archiver server, and 1 monitor server exposed through port number 3000. You can inspect the nodes via the `shardus pm2 list` command.

For usage instructions and available options for the shardus command-line tool, run `shardus --help`.

## Stopping the Network and Cleanup

```bash
shardus stop && shardus clean && rm -rf instances
```
This command will stop the running network, clean up the associated resources, and remove the `instances` folder from the system.

> ## Testing and Metamask

Reviewer: We connect to metamask via the JSON RPC Server which we have not yet started at this point. I would recommend moving this section below the JSON RPC Server setup. That way, we have a running jrpc server we can connect to metamask.

### Install Metamask

Click [here](https://metamask.io/download/) to install the MetaMask extension on your browser.

### Adding Shardeum Network to Metamask

1. Open MetaMask and click the list of networks at the top, then select "Add Network".
2. Add Shardeum automatically or manually:

| Field                | Details                |
| -------------------- | ---------------------- |
| Network Name         | Shardeum Sphinx        |
| New RPC URL          | <http://localhost:8080> |
| Chain ID             | 8082                   |
| Currency Symbol      | SHM                    |
| Block Explorer URL   | <http://localhost:6001/> |

> ## Starting the JSON-RPC Server

Reviewer: Consider adding a `Prerequisite` sub section that says, "Install node-gyp". If they run the npm install command without first installing node-gyp, it would fail.

> A Shardeum validator node does not have an integrated RPC API, instead, we provide an Ethereum compatible JSON-RPC server that can be used to interact with the network, allowing the existing Ethereum tools to use Shardeum seamlessly. To start the JSON-RPC server, clone the repository:

Reviewer: ```suggestion
Shardeum validator nodes do not have an integrated RPC API, instead, we provide an Ethereum compatible JSON-RPC server that can be used to interact with the network. This ensures that existing Ethereum tools and smart contracts can interact with the Shardeum network seamlessly. To start the JSON-RPC server, clone the repository:
```

```bash
git clone https://github.com/shardeum/json-rpc-server.git
```

Change directory to `json-rpc-server`, switch to the `dev` branch, and install the required dependencies:

```bash
cd json-rpc-server
git switch dev
npm install
```

If you want to modify the chainId or the port number, go to the `src/config.ts` file:

```typescript
chainId: 8082,
port: 8080
```

The RPC URL for using Metamask with Remix IDE and for running scripts is `http://localhost:port` (default: <http://localhost:8080>).

### Run the JSON-RPC Server

To run the server, use the command:

```bash
npm run start
```

You now have a fully functioning Shardeum network on your local machine! 🎉

> ## Additional Features and Configurations

Reviewer: For this section, I think we can do without it for now to keep the docs simple. However, if you feel convinced that it is a good addition to have now, then consider adding a bit more information to make it very useful to readers.

For instance, 
- Why should they reconfigure the state manager? 
- What does setting `executeInOneShard` to `true` do for them? etc.

That way, if anyone is choosing to make these configs, they know exactly why and what to expect.
